spring-cloud-stream-binder-jms-common/src/main/java/org/springframework/cloud/stream/binder/jms/utils/SpecCompliantJmsHeaderMapper.java

```diff
@@ -0,0 +1,42 @@
+package org.springframework.cloud.stream.binder.jms.utils;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.jms.Message;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.integration.jms.DefaultJmsHeaderMapper;
+import org.springframework.messaging.MessageHeaders;
+
+/**
+ * Replace all header names that contain '-' with '_' due to IBM MQ header name
+ * constraints.
+ *
+ * See http://stackoverflow.com/a/30024766/2408961 for context.
+ *
+ * @author Donovan Muller
+ */
+public class SpecCompliantJmsHeaderMapper extends DefaultJmsHeaderMapper {
```

Comment:
I'm not sure that such a mutation should be a part of Framework.
In fact the header `my_foo` isn't like `my-foo`. They are fully different identificators.

I may won't mind to add a filtering for the headers with the `-`, but would it honor all other JMS specs?
Recently we closed as `Won't Fix` JIRA about similar problem for headers with the `.`, which don't work with  IBM MQ: https://jira.spring.io/browse/INT-4205.

So, IMO this rewriting logic can be done as a part of target application logic, but not directly in the Framework.
And I'm not sure that I always what to change  `-` to `_`, if you will insist that we can expose such a functionality...